concurrency/RayoPlugin.java

```diff
@@ -0,0 +1,360 @@
+            final String sId = JID.escapeNode(jid.toString());
+            rc.setAttachment(sId);
+            rc.setFrom(jid, component);
+            rc.setCrypto(handset);
+            channels.put(sId, rc);
+
+        } catch (IOException e) {
+            Log.error("Could Not Create Channel.", e);
+        }
+
+        return rc;
+    }
+
+    public void removeChannel(final RelayChannel c) {
+        channels.remove(c.getAttachment());
+        c.close();
+    }
+
+    public String getServiceName() {
+        return serviceName;
+    }
+
+    public void destroyPlugin() {
+        try {
+            componentManager.removeComponent(serviceName);
+            bridge.appStop();
+            SessionEventDispatcher.removeListener(this);
+        } catch (ComponentException e) {
+            Log.error("Could NOT Remove " + serviceName + " Component");
+        }
+        closeAllChannels();
+        executor.shutdownNow();
+        component.doStop();
+    }
+
+    public boolean hasPublicIP() {
+        return hasPublicIP;
+    }
+
+    public int getActiveChannelCount() {
+        return channels.size();
+    }
+
+    /**
+     * Checks whether we have folder with extracted natives, if missing
+     * find the appropriate jar file and extract them. Normally this is
+     * done once when plugin is installed or updated.
+     * If folder with natives exist add it to the java.library.path so
+     * rayo can use those native libs.
+     */
+    private void checkNatives(File pluginDirectory)
+    {
+        // Find the root path of the class that will be our plugin lib folder.
+        try
+        {
+            String nativeLibsJarPath = pluginDirectory.getAbsolutePath() + File.separator + "lib";
+            Log.info("checkNatives." + nativeLibsJarPath);
+
+            File nativeLibFolder = new File(nativeLibsJarPath, "native");
+
+            if(!nativeLibFolder.exists())
+            {
+                nativeLibFolder.mkdirs();
+
+                // lets find the appropriate jar file to extract and
+                // extract it
+                String jarFileSuffix = null;
+                if(OSUtils.IS_LINUX32)
+                {
+                    jarFileSuffix = "rayo-native-linux-32.jar";
+                }
+                else if(OSUtils.IS_LINUX64)
+                {
+                    jarFileSuffix = "rayo-native-linux-64.jar";
+                }
+                else if(OSUtils.IS_WINDOWS32)
+                {
+                    jarFileSuffix = "rayo-native-windows-32.jar";
+                }
+                else if(OSUtils.IS_WINDOWS64)
+                {
+                    jarFileSuffix = "rayo-native-windows-64.jar";
+                }
+                else if(OSUtils.IS_MAC)
+                {
+                    jarFileSuffix = "rayo-native-macosx.jar";
+                }
+
+                JarFile jar = new JarFile(nativeLibsJarPath + File.separator + jarFileSuffix);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *jar*. Currently, there are execution paths that do not contain closure statements. Either a) close *jar* in a try-finally block or b) close the resource by declaring *jar* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).